musig/README.md

```diff
@@ -34,19 +34,63 @@ Ristretto base point in compressed form:
 B = e2f2ae0a6abc4e71a884a961c500515f58e30b6aa582dd8db6a65945e08d2d76
 ```
 
-### Multikey
+### MusigContext
+
+This is a trait with two functions:
+- `commit(&self, &mut transcript)`: takes a mutable transcript, and commits the internal context to the transcript.
+- `challenge(&self, &verification_key, &mut transcript, &nonce_commitment) -> Scalar`: takes a public key and mutable transcript, and returns the 
+  suitable challenge for that public key. 
+- `get_pubkeys(&self) -> Vec<VerificationKey>`: returns the associate public keys.
+
+### Multikey (implements MusigContext)
 
 Fields:
 - transcript: `Transcript`. All of the pubkeys that the multikey are created from are committed to this transcript. 
 - aggregated_key: `VerificationKey`
+- public_keys: `Vec<VerificationKey`
 
 Functions: 
-- `Multikey::new(...)`: detailed more in [key aggregation](#key-aggregation) section. 
-- `Multikey::factor_for_key(&self, &verification_key)`: computes the `a_i` factor, where `a_i = H(<L>, X_i)`. 
-  `<L>`, or the list of pukeys that go into the aggregated pubkey, has already been committed into `self.transcript`. Therefore, this function simply clones `self.transcript`, commits the verification key (`X_i`) into the transcript 
-    with label "X_i", and then squeezes the challenge scalar `a_i` from the transcript with label "a_i".
+- `Multikey::new(...) -> Self`: detailed more in [key aggregation](#key-aggregation) section. 
+
+- `Multikey::commit(&self, &mut transcript)`: Commits `self.aggregated_key` to the input `transcript` with label "X".
+
+- `Multikey::challenge(&self, &verification_key, &mut transcript, &nonce_commitment) -> Scalar`: 
+  Computes challenge `c_i = a_i * c`, where `a_i = H_agg(<L>, X_i)` and `c = H_sig(X, R, m)`.
+
+  For calculating `a_i`, `<L>` (the list of pubkeys that go into the aggregated pubkey)
+  has already been committed into `self.transcript`. Therefore this function simply clones `self.transcript`, 
+  commits the verification key (`X_i`) into the transcript with label "X_i", 
+  and then squeezes the challenge scalar `a_i` from the transcript with label "a_i".
+
+  For calculating `c`, the message `m` has already been committed to the input `transcript`, 
+  so the function only needs to commit `self.aggregated_key` with label "X" and input `hash_commitment` with label "R", 
```

Comment:
AFAIK, the `MusigContext::commit()` has already committed X, and then (inside Party::receive_commitments) R was committed too, so we only need to squeeze the challenge in this method.

Also: the MusigContext should probably be a private trait, so the usage of these commit/challenge helpers is hardcoded correctly under the public API, so no one is supposed to compose usage of these in their protocols. This is strictly the internal organization of the Musig implementation.

Reply:
added a note that musigcontext should be a private trait.

Also, nowhere in the party transitions was `MusigContext::commit()` being called in the design... only in `MusigContext::challenge()`. So, as I detailed here (https://github.com/interstellar/slingshot/pull/260#discussion_r271080929), the current design and your proposed design are both valid, but different in where the commitment is happening.